Improve custom JSON serialization code sample

brutaldev opened the pull request:
1. Remove pointless constructors, the arguments are not used.
2. Use consistent parameter naming on both MessageBodySerializers (the actual interface `IServiceRemotingResponseMessageBodySerializer` has the wrong name for the `Serialize` method).
3. Wrap `writeStream` and  `jsonWriter` in `using` statements so they get disposed, other classes do this but `ServiceRemotingRequestJsonMessageBodySerializer` did not.
4. Use improved `JsonRemotingResponseBody` and `JsonRemotingRequestBody` classes as proposed by "IvanPfeff" in the comments on https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-remoting. The sample currently breaks with similar types because the type information is lost on the round trip. Even integers do not serialize/deserialize correctly and you get errors like `Unable to cast object of type 'System.Int64' to type 'System.Int32'.` which are difficult to debug. This new code fixes the type stripping problem making the code sample mode robust and reliable.
5. Fixed the code in section 2 to used a named parameter. The position of the `new ServiceRemotingJsonSerializationProvider` argument is not the `serializationProvider` (it's `remotingListenerSettings`) so this does not compile without the named parameter.

azurecla commented:
Hi __@KylieLiang__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

deneha commented:
@KylieLiang Thanks for this contribution.  If an item should not be public, then you need to use the azure-content-pr (private repository).  Also, you did not specify the date for merge.

deneha commented:
@KylieLiang Closing this item.  Please duplicate this PR in the private repository.

brutaldev committed:
Improve custom JSON serialization code sample

1. Remove pointless constructors, the arguments are not used.
2. Use consistent parameter naming on both MessageBodySerializers.
3. Wrap `writeStream` and  `jsonWriter` in `using` statements so they get disposed, other classes do this but ServiceRemotingRequestJsonMessageBodySerializer did not.
4. Use improved `JsonRemotingResponseBody` and `JsonRemotingRequestBody` classes as proposed by "IvanPfeff" in the comments. The same currently breaks all the time with similar types because the type information is lost on the round trip. Even integers do not serialize/deserialize correctly and you get errors like `Unable to cast object of type 'System.Int64' to type 'System.Int32'.` This new code fixes that.
5. Fixed the code in section 2 to used a named parameter. The position of the `ServiceRemotingJsonSerializationProvider` argument is not the `serializationProvider` so this does not compile.

suchiagicha commented on articles/service-fabric/service-fabric-reliable-services-communication-remoting.md:
```diff
@@ -311,50 +301,59 @@ Following example uses Json Serialization with Remoting V2.
         public void Set(object response)
         {
-            this.Value = response;
+            Value = JsonConvert.SerializeObject(response);
```
@brutaldev 
Wont this means double serialization and hence less performant ?  For the types like int , guid,, date time where json doesn't deserialize it properly , you could add custom converter for those.

brutaldev replied:
I don't see the double serialization here, perhaps you meant extra/unnecessary serialization? The `response` object could be a regular type (string, int, long, DateTime etc) or it could be a much more complex custom data type. To improve performance  the type could be checked before going through the serializer/deserializer and avoid if it's cant make the trip without it.

The issue here is that complex types containing things like `long` data type properties incorrectly get serialized as int and then break on the other side with data type conversion errors.

suchiagicha commented on articles/service-fabric/service-fabric-reliable-services-communication-remoting.md:
```diff
@@ -311,50 +301,59 @@ Following example uses Json Serialization with Remoting V2.
+                return null;
+            }
+    
+            return JsonConvert.DeserializeObject(Value, paramType);
```
same coment as above. Double De-Serialization .

suchiagicha commented on articles/service-fabric/service-fabric-reliable-services-communication-remoting.md:
```diff
@@ -219,90 +219,80 @@ Following example uses Json Serialization with Remoting V2.
+                {
+                    TypeNameHandling = TypeNameHandling.All
+                });
+                serializer.Serialize(jsonWriter, serviceRemotingRequestMessageBody);
```
@brutaldev 
This is first time you are serializing the serviceRemotingRequestMessageBody.

brutaldev replied:
OK, I will debug further tomorrow and revert back to you with what I find. Reluctant to make sweeping changes since this is working well in production for us.

suchiagicha commented on articles/service-fabric/service-fabric-reliable-services-communication-remoting.md:
```diff
@@ -311,50 +301,59 @@ Following example uses Json Serialization with Remoting V2.
         public void Set(object response)
         {
-            this.Value = response;
+            Value = JsonConvert.SerializeObject(response);
```
@brutaldev 
This is second serialization of same request.  Issue is not just with response. Its with both ,request and response.
I would say do the serialization at remoting object level (your first serialization) and for second serialization, do casting if needed. You could put the check for the types where the Json doesn't do converstion correctlly. (like long, guid , datetime)  inside the get method. You could also use the custom converter for those types .

brutaldev replied:
When I tested this the response is not a string, so it was not being serialized previously.

brutaldev replied:
Honestly, I think the MS team needs to improve this sample if it's that clearly broken.

brutaldev replied:
After debugging the code more, we are reminded why the double serialization is there. The `Value` property in the `IServiceRemotingResponseMessageBody` implementation is serialized **without type information** which is then serialized again in `IServiceRemotingResponseMessageBodySerializer` implementation, this time **with the type information**.

The reason for this is that if you use the normal response type to be serialized in `Serialize`, then type information is stored in the JSON created for the `Value` property. The payload is orders of magnitude larger than if it is just a regular JSON `string` because it will include ALL properties with default values and ALL class/value type information. For a complex object and chatty services, you will quickly run out of memory from all the string allocations.

Therefore, serializing the `Value` property without type information and excluding properties with default values creates a much smaller payload for complex objects which is worth the CPU overhead vs memory use. Different people's mileage may vary, but in general we serialize arrays of large objects between services so doing it this way is the better of the two evils.

An optimization would be to check for value types and strings and avoid the extra serialization of `Value` since you get no benefit in turning those into JSON. A much greater optimization would be to statically create the `JsonSerializerSettings` object which is a very expensive operation.

The type failures where in parameter serialization because there is no type conversion on line #357 which I'm guessing breaks the dynamic invocation down the line on the receiver side. I was confused by your comment about the "second serialization" when in fact the `IServiceRemotingResponseMessageBody` value object is serialized first and then serialized again in `Serialize` to maintain type information.

This entire sample could be improved all over the place, our current production code doesn't look much like the sample anyway as we have our own optimizations in place to improve comms specific to our environment.

suchiagicha replied:
@brutaldev 
I think your issue is with Request type (IServiceRemotingRequestMessageBody) but your change is for the IServiceRemotingResponseMessageBody .
Lets have a quick call to discuss this PR . Could you schedule meeting with suchia@microsoft.com?
